source/io/Stream.cpp

```diff
@@ -13,132 +13,175 @@
+                InputStream::s_GetLength,
+                InputStream::s_Destroy,
+            };
+
+            InputStream::InputStream(Aws::Crt::Allocator *allocator)
+            {
+                m_allocator = allocator;
+                AWS_ZERO_STRUCT(m_underlying_stream);
 
-    impl->stream->seekg(currentPosition);
+                m_good = true;
+                m_underlying_stream.impl = this;
+                m_underlying_stream.allocator = m_allocator;
+                m_underlying_stream.vtable = &s_vtable;
+            }
 
-    return AWS_OP_SUCCESS;
-}
+            StdIOStreamInputStream::StdIOStreamInputStream(
+                std::shared_ptr<Aws::Crt::Io::IStream> stream,
+                Aws::Crt::Allocator *allocator) noexcept
+                : InputStream(allocator), m_stream(std::move(stream))
+            {
+            }
 
-static void s_aws_input_stream_cpp_destroy(struct aws_input_stream *stream)
-{
-    aws_input_stream_cpp_impl *impl = static_cast<aws_input_stream_cpp_impl *>(stream->impl);
-    impl->stream = nullptr;
-    aws_mem_release(stream->allocator, stream);
-}
+            bool StdIOStreamInputStream::ReadImpl(ByteBuf &buffer) noexcept
+            {
+                // so this blocks, but readsome() doesn't work at all, so this is the best we've got.
+                // if you don't like this, don't use std::input_stream and implement your own version
+                // of Aws::Crt::Io::InputStream.
+                m_stream->read(reinterpret_cast<char *>(buffer.buffer + buffer.len), buffer.capacity - buffer.len);
+                auto read = m_stream->gcount();
+                buffer.len += static_cast<size_t>(read);
+
+                if (read > 0)
+                {
+                    return true;
+                }
 
-static struct aws_input_stream_vtable s_aws_input_stream_cpp_vtable = {s_aws_input_stream_cpp_seek,
-                                                                       s_aws_input_stream_cpp_read,
-                                                                       s_aws_input_stream_cpp_get_status,
-                                                                       s_aws_input_stream_cpp_get_length,
-                                                                       s_aws_input_stream_cpp_destroy};
+                auto status = GetStatusImpl();
 
-namespace Aws
-{
-    namespace Crt
-    {
-        namespace Io
-        {
-            struct aws_input_stream *AwsInputStreamNewCpp(
-                const std::shared_ptr<Aws::Crt::Io::IStream> &stream,
-                Aws::Crt::Allocator *allocator) noexcept
+                return status.is_valid && !status.is_end_of_stream;
+            }
+
+            StreamStatus StdIOStreamInputStream::GetStatusImpl() const noexcept
+            {
+                StreamStatus status;
+                status.is_end_of_stream = m_stream->eof();
+                status.is_valid = *m_stream ? true : false;
```

Comment:
trivial: be explicit `m_stream->operator bool()` or `static_cast<bool>(*m_stream)`